src/coreclr/tools/Common/TypeSystem/IL/UnsafeAccessors.cs

```diff
@@ -0,0 +1,515 @@
+                    && fieldType == fd.FieldType)
+                {
+                    context.TargetField = fd;
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        private static MethodIL GenerateAccessor(ref GenerationContext context)
+        {
+            ILEmitter emit = new ILEmitter();
+            ILCodeStream codeStream = emit.NewCodeStream();
+
+            // Load stub arguments.
+            // When the target is static, the first argument is only
+            // used to look up the target member to access and ignored
+            // during dispatch.
+            int beginIndex = context.IsTargetStatic ? 1 : 0;
+            int stubArgCount = context.Declaration.Signature.Length;
+            for (int i = beginIndex; i < stubArgCount; ++i)
+            {
+                codeStream.EmitLdArg(i);
+            }
+
+            // Provide access to the target member
+            switch (context.Kind)
+            {
+                case UnsafeAccessorKind.Constructor:
+                    Debug.Assert(context.TargetMethod != null);
+                    codeStream.Emit(ILOpcode.newobj, emit.NewToken(context.TargetMethod));
+                    break;
+                case UnsafeAccessorKind.Method:
+                    Debug.Assert(context.TargetMethod != null);
+                    codeStream.Emit(ILOpcode.callvirt, emit.NewToken(context.TargetMethod));
+                    break;
+                case UnsafeAccessorKind.StaticMethod:
+                    Debug.Assert(context.TargetMethod != null);
+                    codeStream.Emit(ILOpcode.call, emit.NewToken(context.TargetMethod));
+                    break;
+                case UnsafeAccessorKind.Field:
+                    Debug.Assert(context.TargetField != null);
+                    codeStream.Emit(ILOpcode.ldflda, emit.NewToken(context.TargetField));
+                    break;
+                case UnsafeAccessorKind.StaticField:
+                    Debug.Assert(context.TargetField != null);
+                    codeStream.Emit(ILOpcode.ldsflda, emit.NewToken(context.TargetField));
+                    break;
+                default:
+                    Debug.Fail("Unknown UnsafeAccessorKind");
+                    break;
+            }
+
+            // Return from the generated stub
+            codeStream.Emit(ILOpcode.ret);
+            return emit.Link(context.Declaration);
+        }
+
+        private static MethodIL GenerateAccessorSpecificFailure(ref GenerationContext context, string name, bool ambiguous)
+        {
+            ILEmitter emit = new ILEmitter();
+            ILCodeStream codeStream = emit.NewCodeStream();
+
+            MethodDesc md;
+            TypeSystemContext typeSysContext = context.Declaration.Context;
+            if (ambiguous)
+            {
+                codeStream.Emit(ILOpcode.ldstr, emit.NewToken("Ambiguity in binding of UnsafeAccessorAttribute."));
```

Comment:
This is hardcoding strings from the compiler that should be taken out of corelib's resources instead (localization of the compiler doesn't need to match the one of corelib).

The pattern we use is what I wrote in https://github.com/dotnet/runtime/pull/86932#discussion_r1223980769. You can add a new ThrowHelper method to corelib that matches what you want to do. It will avoid all this IL to newobj a constructor and throw it. All this can just be `codestream.EmitCallThrowHelper(emit, method.Context.GetHelperEntryPoint("ThrowHelpers", "ThrowXXXXXXX"));` (replace XXXXXX by how you want to call it).

Reply:
I don't think I have access to the helpers here. The helpers file isn't included in the type system assembly, only in the ILCompiler project. I was trying to avoid including new files in the projects. I assumed there was a reason it wasn't included.

Reply:
Include the file from the ILCompiler.Compiler project - we use the EmitCallThrowHelper in a couple files under src/coreclr/tools/Common/TypeSystem/IL and they all seem to be included from the ILCompiler.Compiler project.

Reply:
> You can add a new ThrowHelper method to corelib that matches what you want to do. It will avoid all this IL to newobj a constructor and throw it. All this can just be codestream.EmitCallThrowHelper(emit, method.Context.GetHelperEntryPoint("ThrowHelpers", "ThrowXXXXXXX")); (replace XXXXXX by how you want to call it).

This doesn't work for `ThrowBadImageFormatException()` because it finds the export that takes a `string` argument. See below.

https://github.com/dotnet/runtime/blob/32440e59f7363f9ce8d31ae40523b09de9bccada/src/coreclr/tools/Common/TypeSystem/IL/HelperExtensions.cs#L50

Reply:
Sigh... [`ThrowHelper.cs`](https://github.com/dotnet/runtime/blob/main/src/coreclr/nativeaot/System.Private.TypeLoader/src/Internal/TypeSystem/ThrowHelper.cs) vs [`ThrowHelpers.cs`](https://github.com/dotnet/runtime/blob/main/src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/ThrowHelpers.cs) vs the other [`ThrowHelpers.cs`](https://github.com/dotnet/runtime/blob/main/src/coreclr/nativeaot/Runtime.Base/src/System/ThrowHelpers.cs) (same namespace but different location) and ...

Reply:
Which means this is choking on the one that takes an `ExceptionStringID`.

https://github.com/dotnet/runtime/blob/f8f3b547591574f8a18b2429cb91348a261024f1/src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/ThrowHelpers.cs#L76-L79

Reply:
This is so frustrating. The infrastructure here is very cumbersome.

Reply:
@MichalStrehovsky Done. I added a new API for throwing a `AmbiguousMatchException`. There is also a fair bit of shenanigans going on in [`TypeLoaderExceptionHelper.cs`](https://github.com/dotnet/runtime/blob/main/src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/TypeLoaderExceptionHelper.cs) with some of the `Create*` functions returning an exception instance and others immediately throwing one. I can clean it up if you like, but it will be in a subsequent PR.

Reply:
The infrastructure is complicated because it handles the situation where:

* The type system can be included from both the compiler and the runtime type system (exception thrown is one of the TypeSystemException descendants vs a proper TypeLoadException)
* The exception can be thrown within the compiler, the message reported to the user (using the localization of the compiler), then caught and a throwing IL generated that will be rethrown _at runtime using the localization of the CoreLib_ (that may be different from the localization of the compiler)

There's a reason I wrote "I wouldn't bother with the exception message." :).

This didn't really need to deal with the string ID because we're not doing the throw/report/generate throwing IL thing - we could have just created a `ThrowUnsafeAccessorAmbiguousMatch` that hardcodes the message and generate a call to that directly (instead of creating a general-purpose one that takes a string ILD). But the exception string ID is fine.

Reply:
> The infrastructure is complicated because it handles the situation where:

That makes sense. It wasn't the part I was referring to.

> There's a reason I wrote "I wouldn't bother with the exception message." :).

Adding a new string wasn't an issue. Although I will say it feels like we've regressed a bit considering coreclr can reuse error strings defined in SPCL, but NAOT needs to redefine the same error strings.